### DISCUSSION TITLE
Modify `build.Instance` before building it?

### DISCUSSION THREAD

**Discussion by cueckoo** [cue-lang/cue#440] [General] [answered]
_Originally opened by @shykes in https://github.com/cuelang/cue/discussions/440_

TLDR: I need to combine the import support of `load.Instances` and the flexibility of `cue.Compile`. Can I alter `build.Instance.Files` to achieve that?

In theory one can produce their own `cuelang.org/go/cue/build.Instance` without relying on `cuelang.org/go/cue/load.Instances`. But in practice that is difficult, because proper handling of external imports all happens in undocumented private functions inside `load`. So I use `load.Instances`... But it is flexible enough for my needs.

Specifically I would like to dynamically change the “root key” for some of the files in the build, so that they are merged at a key different than what their source file says. I need to honor the imports in those files, so I can’t use `cue.Compile`.

I found the following fields in the `cue/build.Instancd` struct:

- `Files`
- `BuildFiles`
- `IncludeFiles`

Of those, the comments tell me that only `BuildFiles` will exist eventually. But for now all three are used. From reading the docs and the source, it’s not entirely clear how.

I am hoping to modify these fields to transform the correct fioes in the correct way. My best bet seems to be modifying the contents of `Files` using the `ast` package. Does that seem 1) possible and 2) a good idea?

Thanks!

**Comment by cueckoo** [comment, marked as answer]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/440#discussioncomment-34545_

You can modify `build.Instance` files, like adding additional `Files`, at will. The compiler will only look at `Files`. 

It is also fine to modify the contents of existing `Files`.

You do need to ensure the Imports maps is updated accordingly to ensured imported packages are loaded as well if the added files introduce any imports, I believe. I may remove this requirement in the future, but for now it can't hurt to add it.

More info:
The building goes in two stages: first it collects the set of files, then it converts those to syntax trees in `Files`. 

By the time `load` is done, all is converted to `Files`, so it safe to ignore any of the other file fields.

**Comment by cueckoo** [comment]
_Original reply by @shykes in https://github.com/cuelang/cue/discussions/440#discussioncomment-35365_

Thank you. Does this mean that some of the fields of `build.Instance` are specific to `load`, and will be ignored by `cue.Build`? If so, that means I could also ignore those fields when producing my own custom `build.Instance`, correct?

**Reply by cueckoo** [reply]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/440#discussioncomment-35365_

Indeed. Most of the fields are just informative and are ignored by `cue.Build`.

I believe the only thing that is being used is `Files` and `Imports`. The latter could really also be eliminated as a requirement.
